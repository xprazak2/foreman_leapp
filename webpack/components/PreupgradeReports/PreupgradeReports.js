import React, { useState } from 'react';
import MessageBox from 'foremanReact/components/common/MessageBox';
import PropTypes from 'prop-types';
import { Button, Row, Spinner } from 'patternfly-react';

import { translate as __ } from 'foremanReact/common/I18n';

import './PreupgradeReports.scss';

import PreupgradeReportsList from '../PreupgradeReportsList';
import FixAlert from './components/FixAlert';

import {
  invocationFactory,
  flattenEntries,
  isEmpty,
  anyEntriesFixable,
  idsForInvocationFromReports,
  idsForInvocationFromEntries,
} from './PreupgradeReportsHelpers';

import { PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_SUCCESS } from '../../consts';

const PreupgradeReports = ({
  preupgradeReports,
  loading,
  error,
  onFixEntries,
  fixAllWorking,
  fixAllError,
}) => {
  if (!isEmpty(error)) {
    return (
      <MessageBox
        key="preupgrade-reports-error"
        icontype="error-circle-o"
        msg={`Could not retrieve data: ${error.statusText} - ${error.errorMsg}`}
      />
    );
  }

  const [checked, setChecked] = useState([]);

  const isSelected = entry => checked.some(item => item.id === entry.id);

  const anySelected = checked.length > 0;

  const toggleSelected = (entry, isEntrySelected) => {
    if (isEntrySelected) {
      setChecked(checked.filter(item => item.id !== entry.id));
    } else {
      setChecked([entry, ...checked]);
    }
  };

  const onFixAll = reports => {
    const ids = idsForInvocationFromReports(reports);

    onFix(ids);
  };

  const onFixSelected = () => {
    const ids = idsForInvocationFromEntries(checked);

    onFix(ids);
  };

  const onFix = async ids => {
    const action = await onFixEntries(invocationFactory(ids));

    if (action.type === PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_SUCCESS) {
      window.location.pathname = `/job_invocations/${action.payload.id}`;
    }
  };

  return (
    <React.Fragment>
      <Row>
        <div className="title-filter col-md-4">&nbsp;</div>
        <div id="title_action" className="col-md-8">
          <div className="btn-toolbar pull-right">
            {fixAllWorking && (
              <div id="toolbar-spinner">
                <Spinner loading size="sm" />
              </div>
            )}
            <Button
              onClick={() => onFixSelected()}
              disabled={!anyEntriesFixable(preupgradeReports) || fixAllWorking || !anySelected}
            >
              {__('Fix Selected')}
            </Button>
            <Button
              onClick={() => onFixAll(preupgradeReports)}
              disabled={!anyEntriesFixable(preupgradeReports) || fixAllWorking}
            >
              {__('Fix All')}
            </Button>
          </div>
        </div>
      </Row>
      {!isEmpty(fixAllError) && <FixAlert message={fixAllError.message} />}
      <PreupgradeReportsList
        allEntries={flattenEntries(preupgradeReports)}
        fixAllWorking={fixAllWorking}
        isSelected={isSelected}
        toggleSelected={toggleSelected}
      />
    </React.Fragment>
  );
};

PreupgradeReports.propTypes = {
  preupgradeReports: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.object.isRequired,
  onFixEntries: PropTypes.func.isRequired,
  fixAllWorking: PropTypes.bool.isRequired,
  fixAllError: PropTypes.object.isRequired,
};

export default PreupgradeReports;
