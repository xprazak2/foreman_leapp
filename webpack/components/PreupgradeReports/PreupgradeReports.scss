@import '~@theforeman/vendor/scss/variables';

#leapp_preupgrade_report {
  .list-view-pf.working {
    .list-group-item.list-view-pf-stacked {
      color: $color-pf-black-400;

      &:hover{
        background-color: $color-pf-white;
        border-left-color: $color-pf-white;
        border-right-color: $color-pf-white;
      }
    }
  }
}
