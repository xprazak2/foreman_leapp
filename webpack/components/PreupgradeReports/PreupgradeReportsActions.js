import api from 'foremanReact/API';
import { deepPropsToCamelCase } from 'foremanReact/common/helpers';

import {
  PREUPGRADE_REPORTS_REQUEST,
  PREUPGRADE_REPORTS_SUCCESS,
  PREUPGRADE_REPORTS_FAILURE,
} from '../../consts';

export const getPreupgradeReports = url => async dispatch => {
  dispatch({ type: PREUPGRADE_REPORTS_REQUEST });

  try {
    const { data } = await api.get(url);
    return dispatch({
      type: PREUPGRADE_REPORTS_SUCCESS,
      payload: deepPropsToCamelCase(data),
    });
  } catch (error) {
    return dispatch(errorHandler(PREUPGRADE_REPORTS_FAILURE, error));
  }
};

const errorHandler = (msg, err) => {
  const error = {
    errorMsg: 'Failed to fetch preupgrade reports from server.',
    statusText: err.response.statusText,
  };
  return { type: msg, payload: { error } };
};
