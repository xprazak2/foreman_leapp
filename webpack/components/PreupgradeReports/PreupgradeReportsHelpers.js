export const flattenEntries = reports =>
  reports.reduce((memo, report) => [...memo, ...report.entries], []);

const entryWithFixKind = kind => entry =>
  entry.detail &&
  entry.detail.remediations &&
  entry.detail.remediations.some(remediation => remediation.type === kind);

export const entryFixable = entryWithFixKind('command');

export const isInhibitor = entry => entry.flags && entry.flags.some(flag => flag === 'inhibitor');

export const byInhibitor = value => entry => {
  const inhibitsUpgrade = isInhibitor(entry);

  if (value === 'yes') {
    return inhibitsUpgrade;
  }

  if (value === 'no') {
    return !inhibitsUpgrade;
  }

  return true;
};

export const isEmpty = obj => Object.keys(obj).length === 0;

export const anyEntriesFixable = reports =>
  flattenEntries(reports).some(entryFixable);

export const idsForInvocation = reports =>
  reports.reduce(
    (memo, report) => {
      report.entries.forEach(entry => {
        if (entryFixable(entry)) {
          memo.entryIds = [...memo.entryIds, entry.id];

          if (!memo.hostIds.includes(report.hostId)) {
            memo.hostIds = [...memo.hostIds, report.hostId];
          }
        }
      });
      return memo;
    },
    { hostIds: [], entryIds: [] }
  );

export const entriesPage = (entries, pagination) => {
  const offset = (pagination.page - 1) * pagination.perPage;

  return entries.slice(offset, offset + pagination.perPage);
};

export const filterEntries = (attribute, value, entries) => {
  if (!value) {
    return entries;
  }

  if (attribute === 'fix') {
    return entries.filter(entryWithFixKind(value));
  }

  if (attribute === 'inhibitor') {
    return entries.filter(byInhibitor(value))
  }

  return entries.filter(entry =>
    entry[attribute].toLowerCase().includes(value.toLowerCase())
  );
};
