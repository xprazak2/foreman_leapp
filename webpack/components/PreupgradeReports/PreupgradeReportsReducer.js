import Immutable from 'seamless-immutable';

import {
  PREUPGRADE_REPORTS_REQUEST,
  PREUPGRADE_REPORTS_SUCCESS,
  PREUPGRADE_REPORTS_FAILURE,
  PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_REQUEST,
  PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_SUCCESS,
  PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_FAILURE,
} from '../../consts';

export const initialState = Immutable({
  loadingPreupgradeReports: false,
  preupgradeReports: [],
  error: {},
  fixAllWorking: false,
  fixAllError: {},
});

const reducer = (state = initialState, action) => {
  const { payload } = action;
  switch (action.type) {
    case PREUPGRADE_REPORTS_REQUEST:
      return state.set('loadingPreupgradeReports', true);
    case PREUPGRADE_REPORTS_SUCCESS:
      return state.merge({
        loadingPreupgradeReports: false,
        preupgradeReports: payload,
      });
    case PREUPGRADE_REPORTS_FAILURE:
      return state.merge({
        error: payload.error,
        loadingPreupgradeReports: false,
      });
    case PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_REQUEST:
      return state.set('fixAllWorking', true);
    case PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_SUCCESS:
      return state.set('fixAllWorking', false);
    case PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_FAILURE:
      return state.merge({
        fixAllWorking: false,
        fixAllError: { ...payload }, // what is payload???!
      });
    default:
      return state;
  }
};

export default reducer;
