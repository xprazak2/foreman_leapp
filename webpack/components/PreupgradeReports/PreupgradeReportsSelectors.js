export const selectPreupgrade = state => state.foremanLeapp.preupgrade;

export const selectPreupgradeReports = state => selectPreupgrade(state).preupgradeReports;
export const selectLoadingPreupgradeReports = state =>
  selectPreupgrade(state).loadingPreupgradeReports;
export const selectError = state => selectPreupgrade(state).error;
