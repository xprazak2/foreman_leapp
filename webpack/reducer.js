import { combineReducers } from 'redux';

import preupgrade from './components/PreupgradeReports/PreupgradeReportsReducer';

export default combineReducers({ preupgrade });
