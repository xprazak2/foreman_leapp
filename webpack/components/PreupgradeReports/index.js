import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';

import PreupgradeReports from './PreupgradeReports';

import {
  selectPreupgradeReports,
  selectLoadingPreupgradeReports,
  selectError,
} from './PreupgradeReportsSelectors';

import { getPreupgradeReports, postFixAll } from './PreupgradeReportsActions';

const WrappedPreupgradeReports = ({ url, jobInvocationsUrl }) => {
  const loading = useSelector(state => selectLoadingPreupgradeReports(state));
  const preupgradeReports = useSelector(state =>
    selectPreupgradeReports(state)
  );
  const error = useSelector(state => selectError(state));

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getPreupgradeReports(url));
  }, [url]);

  const onFixEntries = (postData) => dispatch(postFixEntries(jobInvocationsUrl, postData));

  return (
    <PreupgradeReports
      preupgradeReports={preupgradeReports}
      error={error}
      loading={loading}
      onFixEntries={onFixEntries}
    />
  );
};

WrappedPreupgradeReports.propTypes = {
  url: PropTypes.string.isRequired,
};

export default WrappedPreupgradeReports;
