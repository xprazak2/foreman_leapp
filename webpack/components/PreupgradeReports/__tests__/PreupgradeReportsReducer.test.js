import { testReducerSnapshotWithFixtures } from '@theforeman/test';

import reducer, { initialState } from '../PreupgradeReportsReducer';

import {
  PREUPGRADE_REPORTS_REQUEST,
  PREUPGRADE_REPORTS_SUCCESS,
  PREUPGRADE_REPORTS_FAILURE,
} from '../../../consts';

import { preupgradeReports } from './PreupgradeReports.fixtures';

const fixtures = {
  'should return initial state': {
    state: initialState,
    action: {
      type: undefined,
      payload: [],
    },
  },
  'should start loading reports on request': {
    state: initialState,
    action: {
      type: PREUPGRADE_REPORTS_REQUEST,
    },
  },
  'should stop loading on reports success': {
    state: initialState.set('loadingPreupgradeReports', true),
    action: {
      type: PREUPGRADE_REPORTS_SUCCESS,
      payload: preupgradeReports,
    },
  },
  'should show error on request failure': {
    state: initialState.set('loadingPreupgradeReports', true),
    action: {
      type: PREUPGRADE_REPORTS_FAILURE,
      payload: { error: { errorMsg: 'This is error', statusText: 'Not good' } },
    },
  },
};

describe('PreupgradeReportsReducer', () =>
  testReducerSnapshotWithFixtures(reducer, fixtures));
