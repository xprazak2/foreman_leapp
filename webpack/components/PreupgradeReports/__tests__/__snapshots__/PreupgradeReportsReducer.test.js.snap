// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreupgradeReportsReducer should return initial state 1`] = `
Object {
  "error": Object {},
  "loadingPreupgradeReports": false,
  "preupgradeReports": Array [],
  "reportsExpected": false,
}
`;

exports[`PreupgradeReportsReducer should show error on request failure 1`] = `
Object {
  "error": Object {
    "errorMsg": "This is error",
    "statusText": "Not good",
  },
  "loadingPreupgradeReports": false,
  "preupgradeReports": Array [],
  "reportsExpected": false,
}
`;

exports[`PreupgradeReportsReducer should start loading reports on request 1`] = `
Object {
  "error": Object {},
  "loadingPreupgradeReports": true,
  "preupgradeReports": Array [],
  "reportsExpected": false,
}
`;

exports[`PreupgradeReportsReducer should stop loading on reports success 1`] = `
Object {
  "error": Object {},
  "loadingPreupgradeReports": false,
  "preupgradeReports": Array [
    Object {
      "entries": Array [
        Object {
          "flags": Array [],
          "hostname": "host.example.com",
          "id": 42,
          "severity": "Too severe to talk about",
          "title": "Fix me!",
        },
        Object {
          "flags": Array [],
          "hostname": "host.example.com",
          "id": 43,
          "severity": "medium",
          "title": "I am broken too",
        },
      ],
      "hostId": 5,
    },
    Object {
      "entries": Array [
        Object {
          "flags": Array [
            "inhibitor",
          ],
          "hostname": "foo.example.com",
          "id": 44,
          "severity": "high",
          "title": "Octocat is not happy",
        },
        Object {
          "flags": Array [],
          "hostname": "foo.example.com",
          "id": 45,
          "severity": "low",
          "title": "Not enough credits",
        },
        Object {
          "flags": Array [],
          "hostname": "foo.example.com",
          "id": 46,
          "severity": "medium",
          "title": "SELinux is turned off",
        },
        Object {
          "flags": Array [],
          "hostname": "foo.example.com",
          "id": 47,
          "severity": "medium",
          "title": "Root password is too short",
        },
        Object {
          "flags": Array [],
          "hostname": "foo.example.com",
          "id": 49,
          "severity": "high",
          "title": "No chocolate chip cookies in cookie jar",
        },
      ],
      "hostId": 6,
    },
  ],
  "reportsExpected": true,
}
`;
